src/librustc/hir/mod.rs

```diff
@@ -340,12 +340,16 @@ impl PathSegment {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum PathParam {
+    Lifetime(Lifetime),
+    Type(P<Ty>),
+}
+
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
 pub struct PathParameters {
-    /// The lifetime parameters for this path segment.
-    pub lifetimes: HirVec<Lifetime>,
-    /// The type parameters for this path segment, if present.
-    pub types: HirVec<P<Ty>>,
+    /// The generic parameters for this path segment.
+    pub parameters: HirVec<PathParam>,
     /// Bindings (equality constraints) on associated types, if present.
     /// E.g., `Foo<A=Bar>`.
     pub bindings: HirVec<TypeBinding>,
```

Comment:
@eddyb 
I wonder if `bindings` should be included to the list as well.
On one hand, if the goal is to employ exhaustive matching to avoid errors, then they should be included. I would certainly expect a visitor method like `fn visit_generic_arg` (aka `fn visit_path_param`) to visit them as well.
On the other hand they are not quite like other generic arguments and don't directly mirror anything from `GenericParams`.

Reply:
I've seen ICEs caused by associated type bindings being forgotten about (fixed some of them in https://github.com/rust-lang/rust/pull/43532).

Reply:
Yeah, I was considering this too. I think @eddyb's opinion was that they were different enough that it was better keeping them separate, but it'd be nice to have some way to ensure they were always handled too.